src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
+            polygon = self.df.at[node, "geometry"]
+            buff_poly = self.df.at[node, "buff_polygon"]
+            # Query rtree for all polygons within `max_travel_distance` of the original
+            poss_neighbours = list(self.rtree.intersection(buff_poly.bounds))
+            for nbr in poss_neighbours:
+                # If a node is not a habitat class node, don't add the edge
+                if hgraph.nodes[nbr]["class_label"] not in valid_classes:
+                    continue
+                # Otherwise add the edge with distance attribute
+                nbr_polygon = self.df.at[nbr, "geometry"]
+                if not hgraph.has_edge(node, nbr) and nbr_polygon.intersects(buff_poly):
+                    # Since intersects is cheaper than distance, we can save time here
+                    if nbr_polygon.intersects(polygon):
+                        distance = 0.0
+                    else:
+                        distance = polygon.distance(nbr_polygon)
+                    hgraph.add_edge(node, nbr, distance=distance)
+
+        # Remove expanded polygons from df
+        self.df = self.df.drop(columns=["buff_polygon"])
+        # Remove non-habitat nodes from habitat graph
+        hgraph.remove_nodes_from(invalid_indexes)
+        # Add habitat graph to habitats dict
+        self._habitats[name] = hgraph
+
+        print(
+            f"Habitat successfully loaded with {hgraph.number_of_nodes()} nodes",
+            f"and {hgraph.number_of_edges()} edges.",
         )
+
+    def get_graph_components(
+        self, graph: nx.Graph
+    ) -> Tuple[gpd.GeoDataFrame, List[set]]:
+        """Return a GeoDataFrame with graph components.
+
+        This method takes an nx.Graph and determines the individual disconnected
+        graph components that make up the graph. Each row of the returned
+        GeoDataFrame corresponds to a graph component, with entries in column
+        'geometry' being the union of all individual polygons making up a
+        component. Warning: this function can only be passed graphs which are a
+        subgraph of this class's main `graph` attribute, such as the habitat
+        subgraph or the main graph itself.
+
+        This method allows for the UI to visualise components and output their
+        number as a metric. Warning: this method is very slow if the graph
+        consists mostly of one big component, because taking the union is
+        expensive.
+
+        More info on the definition of graph components can be found here:
+        https://en.wikipedia.org/wiki/Component_(graph_theory)
+
+        Args:
+            graph (nx.Graph): nx.Graph object. This must be a subgraph of this
+            class's main `graph` attribute.
+
+        Returns:
+            Tuple: A tuple containing the resulting GeoDataFrame and the list of
+            graph components.
+        """
+        components = list(nx.connected_components(graph))
```

Comment:
Is `components` a list of lists of node indices?

Reply:
It is a list of sets of node indices.